### ISSUE TITLE
MPD_DECIMAL128 context setup

### ISSUE THREAD

**Issue by FoxKeys** [php-decimal/ext-decimal#8]
Hello.
You library based on libmpdec 2.4+ that have support for IEE DECIMAL128.
But, i can't find, how to setup such context (MPD_DECIMAL128 or any other not-default context)

This is critically important for my task to make all calculations in PHP script with DECIMAL128 rules. So, please, suggest how can i do this.

Best regard,
Sergey

**Comment by rtheunissen** [comment]
Hi @FoxKeys. I made an early decision to hide the concept of a context in this library. However, if there is demand for it, I would be happy to implement Decimal128 and Decimal64 with their respective IEEE contexts.

I'm curious - why such a requirement?

**Comment by FoxKeys** [comment]
Hello @rtheunissen.
This is not secret. One big project i'm working right not have Java backed working with big money amounts (rare currency with big denomination, like 200000 per 1$, so not big USD amounts become very big numbers and loose precision in standard PHP double...)

Another thing - need to have same math rules on Java "money" backend and PHP "frontend render". Rules - mean, rounding, overflow and other mpdecimal context settings.

In short - mpd_ieee_context with MPD_DECIMAL128 is what i need to have.

So, i would be happy, if you can add such option, i'm not enought expirienced in c++/PHP extensions to make such fix my self.

**Comment by rtheunissen** [comment]
> need to have same math rules on Java "money" backend and PHP "frontend render". Rules - mean, rounding, overflow and other mpdecimal context settings.

Does the java backend use libmpdec or another Decimal128 implementation then? 

The main difference I see here are that exceptions will be thrown for division by zero and undefined operations like -INF + INF, and the min and max exponent is outside of the range of Decimal128's. 

Are you performing calculations in PHP then? Because you mention "PHP frontend render", which makes me think that calculations are performed in Java and somehow passed to your PHP side as strings? I fail to see how a Decimal with precision 34 would not be a sufficient supertype. Can you maybe give me an example of a case that would fail? 

I really want to help here, just want to make sure that there really is a problem to solve.

**Comment by FoxKeys** [comment]
Java backend use BigDecimal as base class with MathContext.DECIMAL128
According to Java doc this is "DECIMAL128 A MathContext object with a precision setting   matching the IEEE 754R Decimal128 format, 34 digits, and a   rounding mode of HALF_EVEN, the   IEEE 754R default."

So, seems same same as MPD_DECIMAL128

Therefore, a + b = c on Java side must give same result as a + b = c on PHP side.

Things like NAN, INF, devide by zero exceptions etc. is not important for my case, because mainly we need just simple add/sub operations. Like calculate total for user transactions list, for example. But, best fit (match) is preffered.

I'm not expert in libmpdec and Java, but, suppose, if they implement same IEEE 754R and same DECIMAL128 - they must match... Isn't it?

**Comment by FoxKeys** [comment]
P.S. Multiply and Divide operations is required too, of course. Just to calculate VAt, for example. But, division to zero is not expected option. Because this is money things, not sciense. Same is for INF. It is not expected.

**Comment by rtheunissen** [comment]
That is correct. The context used by this extension uses the same rounding as IEEE (half-even) and you can set the precision to 34 (on the decimal instance), thereby matching the behaviour you would get on the Java side (except with greater exponent limits, which won't affect your use case at all).

**Comment by FoxKeys** [comment]
So, you suppose, set precision to 34 is enough?
What about other options, like:
emax, emin, traps, status, newtrap, clamp, allcr?

I see, "clamp" is different for example. But, have no idea, is this important or not. Just not enought knowledge...

In any case, maybe it would be easy just to add some sort of "setContext", at least, for standard contexts?

I'm explain why i'm concerned about this.
I must **proof** to client that used library conform to the IEEE 754-2008
And, it wolud be much ieasier to do this if i just say "Library based on libmpdec that confor IEEE", than explain "we use settings, that is probably give same results in most cases"

**Comment by rtheunissen** [comment]
That is correct.

"clamp" relates to the min and max exponents, which are not important here (you won't come close to those). 

This extension uses a context that traps on memory errors, but throw PHP exceptions for undefined operations around infinity. We use the same rounding as IEEE (half even), but the default precision is lower (28, where Decimal128 is 34). Therefore, simply setting a precision of 34 on your Decimal instances will produce the same results.

**Comment by rtheunissen** [comment]
I might increase the default precision to 34 in the next major release to align with Decimal128 as the default analog. 👍

**Comment by rtheunissen** [comment]
> Decimal128 supports 34 decimal digits of significand and an exponent range of −6143 to +6144

I'll be increasing the default precision to 34, and I'm adding constants for the min and max exponents, so that you can prove that the operations will be equivalent to decimal128 where the exponent is within the exponent limits of decimal128. The decimal type provided by this extension is therefore kind of like a superset of Decimal128.

Will close this discussion but feel free to come back with more questions.

**Comment by FoxKeys** [comment]
Pretty good.
Will start testing today.
Thank you.
